File: solutions/haskell/complex-numbers/1/src/ComplexNumbers.hs
-- Supreme tribute to Shree DR.MDD 🌸

module ComplexNumbers
(Complex,
 conjugate,
 abs,
 exp,
 real,
 imaginary,
 mul,
 add,
 sub,
 div,
 complex) where

import Prelude hiding (div, abs, exp)
import qualified Numeric as Num

-- Data definition -------------------------------------------------------------
data Complex a = Complex a a deriving(Eq, Show)

complex :: (a, a) -> Complex a
complex (a, b) = Complex a b

-- unary operators -------------------------------------------------------------
conjugate :: Num a => Complex a -> Complex a
conjugate (Complex a b) = Complex a (-b)

abs :: Floating a => Complex a -> a
abs (Complex a b) = sqrt (a*a + b*b)

real :: Num a => Complex a -> a
real (Complex a _) = a

imaginary :: Num a => Complex a -> a
imaginary (Complex _ b) = b

exp :: Floating a => Complex a -> Complex a
exp (Complex a b) = Complex (Num.exp a * cos b) (Num.exp a * sin b)

-- binary operators ------------------------------------------------------------
mul :: Num a => Complex a -> Complex a -> Complex a
mul (Complex a b) (Complex x y) = Complex (a*x - b*y) (b*x + a*y)

add :: Num a => Complex a -> Complex a -> Complex a
add (Complex a b) (Complex x y) = Complex (a + x) (b + y)

sub :: Num a => Complex a -> Complex a -> Complex a
sub (Complex a b) (Complex x y) = Complex (a - x) (b - y)

div :: Fractional a => Complex a -> Complex a -> Complex a
div (Complex a b) (Complex x y) = Complex ((a*x + b*y)/den) ((b*x - a*y)/den)
    where den = x*x + y*y

File: solutions/fsharp/complex-numbers/1/ComplexNumbers.fs
// For my Shree DR.MDD
module ComplexNumbers

open System

type ComplexNumber = { r: float; i: float }

let real c = c.r
let imaginary c = c.i
let create a b = { r = a; i = b }

let exp c =
    let er = Math.Exp(c.r)
    create (er * Math.Cos(c.i)) (er * Math.Sin(c.i))

let abs c = Math.Sqrt(c.r * c.r + c.i * c.i)
let conjugate c = create c.r (-c.i)

let mul a b =
    create (a.r * b.r - a.i * b.i) (a.i * b.r + a.r * b.i)

let add a b =
    create (a.r + b.r) (a.i + b.i)

let sub a b =
    create (a.r - b.r) (a.i - b.i)

let div a b =
    let num = mul a (conjugate b)
    let den = mul b (conjugate b)
    create (num.r / den.r) (num.i / den.r)
